server/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -637,6 +639,12 @@ public void addAllocFailuresResetListenerTo(ClusterService clusterService) {
         });
     }
 
+    public void setHeapUsageSupplier(HeapUsageSupplier heapUsageSupplier) {
+        if (clusterInfoService instanceof InternalClusterInfoService icis) {
+            icis.setHeapUsageSupplier(heapUsageSupplier);
```

Comment:
This is very awkward and error prone (it requires careful coordination of where this mutable state is set. We can avoid this by following another pattern for plugging this in:

* Construct the supplier in ClusterModule (where AllocationService is created) by passing in the PluginsService, using `loadSingletonServiceProvider`.
* Have the implementation take the owning Plugin instance as a single argument. Access any state the implementation needs from that Plugin instance. Eg if you need ClusterService,  have a SetOnce on the plugin to retain the cluster service, and have the implementation grab the cluster service from the plugin at runtime when it needs it.

Reply:
Thanks Ryan, I had a go at implementing this. It makes it a bit tricky to supply a test implementation. Perhaps there's a trick I didn't see, or maybe we just save the integration testing for where the implementation is.

Reply:
We don't have a mechanism for injecting a test version of SPI right now. If that's important, I think it could be done by having the MockPluginsService override loadServiceProviders and filtering out the production one. But that will require some design thought. Alternatively you could use normal plugin interfaces (just make sure they're in packages that have qualified exports to serverless so we aren't opening this up to all plugins).

Reply:
I don't think it's important, I have configured a dummy supplier in the integration test which allows us to assert that one can be configured. I think I'm happy with that and we can revisit if/when we implement in stateful. Thanks for the input.